### PR TITLE
feat: IDP-1766-crittografare-il-pin-della-cie

### PR DESCRIPTION
**Pull request by LarissaASLeite**
#### List of Changes
- Creation PinBlock
- Hash PinBlock

#### How Has This Been Tested?

- Pre-Deploy Test
    - [x] Unit
    - [ ] Integration (Narrow)
- Post-Deploy Test
    - [ ] Isolated Microservice
    - [x] Broader Integration
    - [ ] Acceptance
    - [ ] Performance & Load

#### Types of changes

- [ ] PATCH - Bug fix (backwards compatible bug fixes)
- [x] MINOR - New feature (add functionality in a backwards compatible manner)
- [ ] MAJOR - Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/java/it/gov/pagopa/payment/instrument/model/PaymentInstrumentCode.java [hunk 1]
```diff
@@ -21,6 +21,8 @@ public class PaymentInstrumentCode {
   private String id;
   private String userId;
   private String idpayCode;
+  private String salt;
+  private String secondFactor;
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/instrument/model/PaymentInstrumentCode.java:25]: why are calling it secondFactor and not fakeNIS?

File: src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/EncryptCodeServiceImpl.java [hunk 2]
```diff
@@ -1,18 +1,81 @@
...
+    String pinBlock = calculatePinBlock(secondFactor, code);
+    return createSHA256Digest(pinBlock, salt);
+  }
+
+  private String calculatePinBlock(String secondFactor, String code) {
+    long startTime = System.currentTimeMillis();
+    // Control code length, must be 5
+    try {
+      if (code.length() < 5) {
+        throw new PaymentInstrumentException(400, "Pin length it's not valid");
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/EncryptCodeServiceImpl.java:30]:
```suggestion
        throw new PaymentInstrumentException(400, "Pin length is not valid");
```

File: src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/EncryptCodeServiceImpl.java [hunk 3]
```diff
@@ -1,18 +1,81 @@
...
+      // Combine CODE and SECOND_FACTOR with XOR
+      final byte[] codeBytes = Hex.decodeHex(codeData.toCharArray());
+      final byte[] fakeNIS = secondFactor.getBytes(StandardCharsets.UTF_8);
+      final byte[] xorResult = new byte[64];
+      for (int i = 0; i < 8; i++) {
+        xorResult[i] = (byte) (codeBytes[i] ^ fakeNIS[i]);
+      }
+
+      String pinBlock = Hex.encodeHexString(xorResult);
+      log.info("[{}] pinBlock generated successfully", GENERATE_PIN_BLOCK);
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/EncryptCodeServiceImpl.java:45]: you are logging too much: we are not interested to have an INFO for operation inside each step.
make all inner logs DEBUG and let just one INFO.
eg: one INFO to log the complete pinBlock generation, and a DEBUG to locate create, hash, cipher operations

ps: the performanceLog is already an INFO log

File: src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/PaymentInstrumentCodeServiceImpl.java [hunk 4]
```diff
@@ -48,12 +49,18 @@ public GenerateCodeRespDTO generateCode(String userId, String initiativeId) {
...
 
+    // generate Salt
+    String salt = generateSaltOrSecondFactor(16);
+
+    // generate secondFactor
+    String secondFactor = generateSaltOrSecondFactor(16);
+
     // encrypt clear code
-    String idpayCode = encryptCodeService.encryptIdpayCode(clearCode);
+    String idpayCode = encryptCodeService.encryptIdpayCode(clearCode, secondFactor, salt);
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/PaymentInstrumentCodeServiceImpl.java:59]: let's give more meaningfull names
```suggestion
    String hashedPinBlock= encryptCodeService.encryptIdpayCode(clearCode, secondFactor, salt);
```

File: src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/EncryptCodeServiceImpl.java [hunk 5]
```diff
@@ -1,18 +1,81 @@
...
-  public EncryptCodeServiceImpl() {
-    // TODO to be implemented
-  }
+  public static final String GENERATE_PIN_BLOCK = "GENERATE_PIN_BLOCK";
+  public static final String HASH_PIN_BLOCK = "HASH_PIN_BLOCK";
 
   @Override
-  public String encryptIdpayCode(String code) {
-    return code;
+  public String encryptIdpayCode(String code, String secondFactor, String salt) {
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/EncryptCodeServiceImpl.java:20]: let's use more meaningful names
```suggestion
  public String buildHashedPinBlock(String code, String secondFactor, String salt) {
```

File: src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/PaymentInstrumentCodeServiceImpl.java [hunk 6]
```diff
@@ -99,6 +106,14 @@ public boolean codeStatus(String userId) {
     return idpayCodeEnabled;
   }
 
+  // Generate Salt or Second factor.
+  // TODO Put length like dynamic variable?
+  private String generateSaltOrSecondFactor(int length) {
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/PaymentInstrumentCodeServiceImpl.java:111]:
```suggestion
  private String generateRandomBase64String(int length) {
```

File: src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/PaymentInstrumentCodeServiceImpl.java [hunk 7]
```diff
@@ -99,6 +106,14 @@ public boolean codeStatus(String userId) {
     return idpayCodeEnabled;
   }
 
+  // Generate Salt or Second factor.
+  // TODO Put length like dynamic variable?
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/instrument/service/idpaycode/PaymentInstrumentCodeServiceImpl.java:110]: resolve this TODO before to merge

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/main/java/it/gov/pagopa/payment/instrument/model/PaymentInstrumentCode.java (hunk 1 above)
Comment (LarissaASLeite) [reply src/main/java/it/gov/pagopa/payment/instrument/model/PaymentInstrumentCode.java:25]: to standardize with the ms MIL

**Review by antonioT90 — CHANGES_REQUESTED** [review]

**Review by antonioT90 — APPROVED** [review]

**Review by frtrinca — APPROVED** [review]

**Pull request merged**